### DISCUSSION TITLE
Config Based Form that utilizes useFieldArray and Controller - issues with controlled values

### DISCUSSION THREAD

**Discussion by lsbyerley** [react-hook-form/react-hook-form#2519] [Q&A] [answered]
Hello everyone! I am having issues with controlling the form values of a form that utilizes the useFieldArray hook with Controller components that are passed a render function. I am using a config file based approach to render our forms. Some of the issues are:

- when adding or removing a row, all input data is cleared out
- the name attribute is not making it into the input component even though its being set in the controller
- On submit, all the form values are cleared in the dom, but they exist in the form data object (as you can see in the screenshot below)

<img width="779" alt="Screen Shot 2020-08-07 at 11 40 18 AM" src="https://user-images.githubusercontent.com/3066258/89662953-cfc02880-d8a2-11ea-835c-79d514e37033.png">

I have setup a codesandbox that mimics the config file setup we are using. If anyone has any ideas for solving the issues I am having, please let me know!

https://codesandbox.io/s/ecstatic-maxwell-xyywz?file=/src/App.js

**Comment by moofoo** [comment, marked as answer]
> when adding or removing a row, all input data is cleared out

You need to pass down the fieldArray 'item' and use it to set the defaultValue of your fields in the fieldArray. Remove, move, swap, etc operations unmount/mount fieldArray elements, so they revert to defaultValue again. Item tracks values based on fieldArray position, so using item to the set the defaultValue of fields ensures they hold on to their values after they re-render.

Also, declaring RowItems in the body of FieldArray is an issue. 

https://codesandbox.io/s/hungry-swartz-iyqyf?file=/src/FieldArray.jsx:0-1765

Changing your FieldArray component like this fixes the 'input data clearing out' issues:
```
import React from "react";
import { Box, Grid } from "theme-ui";
import { set, isArray } from "lodash";
import { useFieldArray, useFormContext, Controller } from "react-hook-form";

//Declaring RowItems in the body of FieldArray was the source of most of the issues
const RowItems = ({ children }) => {
  return (
    <Grid gap={4} columns={[1, 3]}>
      {children}
    </Grid>
  );
};

const FieldArray = ({ name, configItems }) => {
  const defaultItem = {};
  if (configItems && isArray(configItems)) {
    configItems.forEach((i) => set(defaultItem, i.name, i.defaultValue || ""));
  }

  const { control } = useFormContext();
  const { fields, append, remove } = useFieldArray({
    control,
    name,
    keyName: "itemId" //you have a component with name 'id', so you need this so it doesn't conflict with item.id
  });

  const handleAdd = () => {
    append(defaultItem);
  };

  const handleRemove = (index) => {
    remove(index);
  };

  return (
    <Box>
      <h2>{name}</h2>
      <hr />
      {fields.map((item, index) => {
        return (
          <RowItems key={item.itemId}>
            {configItems.map((configItem) => {
              return (
                <Controller
                  render={configItem.props.render}
                  key={`${configItem.name}${index}`}
                  name={`${name}[${index}].${configItem.name}`}
                  control={control}
                  defaultValue={item[configItem.name]}
                />
              );
            })}
            <button onClick={() => handleRemove(index)}>Remove</button>
          </RowItems>
        );
      })}
      <button mt={4} type="button" onClick={() => handleAdd()}>
        Add Row
      </button>
    </Box>
  );
};

export default FieldArray;

```

**Reply by lsbyerley** [reply]
Thank you for taking a look! I've updated my codesandbox with your suggestion and yeah it is still a little wonky. Maybe some others can provide some insight

**Comment by moofoo** [comment]
Check out my updated comment. 

Also I'm apparently very bored because I did a little rewrite of your app using some methods from a similar project of my own , (config based form generation), maybe it will be of some help! https://codesandbox.io/s/cranky-diffie-q0x00?file=/src/FieldArray.jsx

**Reply by lsbyerley** [reply]
Haha you must be! Appreciate the new example, it does feel like a somewhat cleaner approach by not having the render() function in the config file. I may have to consider this. I'm assuming it would be easy to add more fields also, like a select dropdown

**Reply by bluebill1049** [reply]
Thanks a lot for your support @moofoo 🙏
